### PR TITLE
issue #118: implement delete_inspection

### PR DESCRIPTION
**Pull request by k-allagbe**
- [x] new schema fertiscan_0.0.13
- [x] delete cascades
- [x] delete_inspection sql function
- [x] datastore.fertiscan delete function
- [x] tests
- [x] docs

### CODE REVIEW

**Comment by k-allagbe** [comment]
On hold until https://github.com/ai-cfia/ailab-datastore/issues/132 is fixed.

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (Francois-Werbrouck) [inline datastore/db/bytebase/FertiScan/delete_inspection_function.sql:34]: Is there instances where 2 organization might ahve the same location? If so, it might cause confloct for the organization information with said location.

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (Francois-Werbrouck) [inline datastore/db/bytebase/FertiScan/delete_inspection_function.sql:47]: This should be a trigger and wouldn't require a select

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (Francois-Werbrouck) [inline datastore/db/bytebase/FertiScan/delete_inspection_function.sql:62]: Why do we need to return a json of the deleted row?

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (Francois-Werbrouck) [inline datastore/db/bytebase/FertiScan/delete_inspection_function.sql:36]: This whole functioon should be only a trigger. Also I dont get what `related data` this is deleting? The comment is missleading

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (Francois-Werbrouck) [inline datastore/db/bytebase/FertiScan/delete_inspection_function.sql:77]: Why isnt this handled with a Cascade Constraint?

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (Francois-Werbrouck) [reply datastore/db/bytebase/FertiScan/delete_inspection_function.sql:36]: I was expecting this function to call and delete the label_information data

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (Francois-Werbrouck) [inline datastore/db/bytebase/FertiScan/delete_inspection_function.sql:99]: Why are we only deleting the org_info and not the rest of the child tables?

File: datastore/db/metadata/inspection/__init__.py
Comment (Francois-Werbrouck) [inline datastore/db/metadata/inspection/__init__.py:98]: We dont need this

File: datastore/db/queries/inspection/__init__.py
Comment (Francois-Werbrouck) [inline datastore/db/queries/inspection/__init__.py:397]: I dont see the value of returning a deleted record. If it is deleted, the data shouldn't be passed around anymore

File: datastore/fertiscan/__init__.py
Comment (Francois-Werbrouck) [inline datastore/fertiscan/__init__.py:304]: I usually manually check if the user_id is associated with the inspection before calling the function to give better insight to our other layers on the errors

File: doc/fertiscan/delete-inspection.md
Comment (Francois-Werbrouck) [inline doc/fertiscan/delete-inspection.md:1]: This helps alot understand how the delete work.

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (Francois-Werbrouck) [reply datastore/db/bytebase/FertiScan/delete_inspection_function.sql:99]: With the doc, I now see the rest is deleted by cascade

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (Francois-Werbrouck) [inline datastore/db/bytebase/FertiScan/delete_inspection_function.sql:1]: I dont see any handling of deleting the picture_set_id

**Review by Francois-Werbrouck — CHANGES_REQUESTED** [review]
I think this needs a bit of tweakign before being ready :)

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/delete_inspection_function.sql:25]: Yes. This is now handled by catching possible foreign key exception and raising a notice.

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/delete_inspection_function.sql:38]: That snippet is to make sure the given user_id is responsible for the inspection. I'm not convinced a trigger is appropriate for such checks. And I'm not sure how I would pass the user_id to the trigger. Please suggest an example.

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/delete_inspection_function.sql:53]: I return the inspection because I use the picture_set_id it contains afterwards to delete the corresponding picture_set (using existing datastore methods). I could of course:
- return the picture_set_id that would be strange given that the centre of attention is the inspection
- fetch the picture_set_id prior to deleting the inspection, but that's one more unnecessary call to the db

Please, suggest an alternative you find this inappropriate.

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/delete_inspection_function.sql:27]: yes, deleting the label_information is a solution but it is strange to delete a label_information when the function is named delete_inspection even though the inspection will eventually be cascade deleted.

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/delete_inspection_function.sql:68]: Sample doesn't have inspection_id. It's the other way around.

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/delete_inspection_function.sql:90]: Please, don't forget to resolve the comments.

File: datastore/db/metadata/inspection/__init__.py
Comment (k-allagbe) [reply datastore/db/metadata/inspection/__init__.py:98]: It's not needed, it's for convenience. I prefer working in the code with models than dictionaries. One advantage: If a data field name changes, it has to be updated everywhere it is used with a dictionary, but only once with a model (vscode's rename feature).

File: datastore/db/queries/inspection/__init__.py
Comment (k-allagbe) [reply datastore/db/queries/inspection/__init__.py:397]: Why: https://github.com/ai-cfia/ailab-datastore/pull/128#discussion_r1741263788

It's fine keeping deleted data in app memory as long as it is known to be so. Because it could contain data still useful elsewhere.

File: datastore/fertiscan/__init__.py
Comment (k-allagbe) [reply datastore/fertiscan/__init__.py:304]: Is it all the same if it is checked inside the function? Checking inside the function avoid an additional round trip to the DB.

File: doc/fertiscan/delete-inspection.md
Comment (k-allagbe) [reply doc/fertiscan/delete-inspection.md:1]: Please, don't forget to resolve the comments.

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/delete_inspection_function.sql:1]: It is handled separately in the datastore function by using `delete_picture_set_permanently`. This is described in the last section of the document.

If some other logic is needed to delete the picture sets in fertiscan, that could make another issue. For now I think this is acceptable.

**Comment by k-allagbe** [comment]
@Francois-Werbrouck 

Please revisit your last change requests, I've answered them. 

How do you suggest we handle the schema version conflict? Because I see there's already a `0.0.13` in bytebase.

**Comment by Francois-Werbrouck** [comment]
> How do you suggest we handle the schema version conflict? Because I see there's already a `0.0.13` in bytebase.

You can create a schema to work on your issue in our database and set the secret to the right value. This will help me with seing the unittests 

>Please revisit your last change requests, I've answered them.

I'm sorry, but could I ask if this could be rebased on main before? I feel like the merge might impact your changes? I will do so right after! 👍🏻

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (Francois-Werbrouck) [reply datastore/db/bytebase/FertiScan/delete_inspection_function.sql:53]: On Nachet we chose:

>fetch the picture_set_id prior to deleting the inspection, but that's one more unnecessary call to the db

Your solution, 
>return the picture_set_id, but that would be strange given that the centre of attention is the inspection

is also viable and require less db transactions. I was scared this data would be sent to other Layers but your answer satisfy me!

File: datastore/fertiscan/__init__.py
Comment (Francois-Werbrouck) [reply datastore/fertiscan/__init__.py:304]: If we check it before hand, we can return a custom error for the BE to have better insight. If you check it within the DB function, I feel like its harder to give context on the error, unless you know an easy way.

**Comment by Francois-Werbrouck** [comment]
I've answered and resolved some comments

File: tests/fertiscan/db/delete_inspection/test_delete_organization_info.py
Comment (Francois-Werbrouck) [inline tests/fertiscan/db/delete_inspection/test_delete_organization_info.py:128]: This could be solved in another issue, but writing queries in tests makes them harder to maintain. I advise creating python function and use them, this makes the maintenance more dynamic and less redundant

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (Francois-Werbrouck) [reply datastore/db/bytebase/FertiScan/delete_inspection_function.sql:25]: I dont see us handling the following use case: 

organization_A uses location_A
organization_B uses location_A

From my understanding, if you delete organization_A, you also delete location_A. 
However you shouldn't since organization_B also uses location_A.

If this case is handled, please point me towards where, because I can't seem to find it 🫤

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (Francois-Werbrouck) [reply datastore/db/bytebase/FertiScan/delete_inspection_function.sql:53]: you can resolve the comment once read

File: datastore/db/metadata/inspection/__init__.py
Comment (Francois-Werbrouck) [reply datastore/db/metadata/inspection/__init__.py:98]: I didn't know about that feature, I might need to learn about it to make our tests more dynamic

File: datastore/db/queries/inspection/__init__.py
Comment (Francois-Werbrouck) [reply datastore/db/queries/inspection/__init__.py:424]: you can resolve once read

File: datastore/fertiscan/__init__.py
Comment (k-allagbe) [reply datastore/fertiscan/__init__.py:304]: The easiest way I found was to raise in `delete_inspection` a `InspectionDeleteError` from the db exception that would be raised when the user_id isn't right, passing along all the details.

Check `"fertiscan_0.0.13".delete_inspection`:
```sql
    IF NOT EXISTS (
        SELECT 1
        FROM "fertiscan_0.0.13".inspection 
        WHERE id = p_inspection_id 
        AND inspector_id = p_inspector_id
    ) THEN
        RAISE EXCEPTION 'Inspector % is not the creator of inspection %', p_inspector_id, p_inspection_id;
    END IF;
```

and `inspection.delete_inspection`
```python
    except (Error, DatabaseError, OperationalError) as e:
        raise InspectionDeleteError(f"Database error occurred: {str(e)}") from e
```

That should ensure the exception details bubble all the way up to the BE.

File: datastore/db/metadata/inspection/__init__.py
Comment (k-allagbe) [reply datastore/db/metadata/inspection/__init__.py:98]: It's akin to renaming a variable name:
image

File: datastore/db/bytebase/FertiScan/delete_inspection_function.sql
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/delete_inspection_function.sql:25]: In the schema, note the fk constraint:
```sql
    CREATE TABLE "fertiscan_0.0.13"."organization_information" (
        "id" uuid PRIMARY KEY DEFAULT uuid_generate_v4(),
        "name" text,
        "website" text,
        "phone_number" text,
        "location_id" uuid REFERENCES "fertiscan_0.0.13".location(id),  <--- here
        "edited" boolean DEFAULT false
    );
```


In the `after_org_info_delete_location_trig` trigger function, note that I raise a simple notice in case there is a fk error:
```sql
CREATE OR REPLACE FUNCTION "fertiscan_0.0.13".after_org_info_delete_location_trig()
RETURNS TRIGGER
LANGUAGE plpgsql
AS $$
BEGIN
    IF OLD.location_id IS NOT NULL THEN
        BEGIN
            DELETE FROM "fertiscan_0.0.13".location
            WHERE id = OLD.location_id;
        EXCEPTION WHEN foreign_key_violation THEN <--- from here
            RAISE NOTICE 'Location % is still referenced by another record and cannot be deleted.', OLD.location_id;
        END;
    END IF;

    RETURN NULL;
END;
$$;

```

So in your use case, if organization_A is deleted, it should not delete location because it is organization_A that has location's id. But even in the advent the location's deletion is attempted, if it is related to any organization_info, a fk exception will be raised. And in the case of `after_org_info_delete_location_trig`, the exception will just be silenced.

File: tests/fertiscan/db/delete_inspection/test_delete_organization_info.py
Comment (k-allagbe) [reply tests/fertiscan/db/delete_inspection/test_delete_organization_info.py:128]: Created https://github.com/ai-cfia/ailab-datastore/issues/158 to track this.

**Review by Francois-Werbrouck — APPROVED** [review]
with the creation of #158 This PR looks good to me 👍🏻

**Pull request merged**
